Code-4-Community/c4c-ops
# Ojn03/feat update user

### ℹ️

Closes #5 

### 📝 Description

Added an endpoint that allows users to update information in the database so that: 
- Only admins can update another person's info
- Members and Alumni cannot update Applicant info
- Applicants can only update their own info

Briefly list the changes made to the code:
1. Added a PUT route in the user controller to accept queries for updateUser and direct logic accordingly. The query accepts userId param as a string and a body as a UpdateUserDTO (defined below)
2. defined a data transfer object class (UpdateUserDTO), to define the expected structure of the body in the put request. 
3. updated user service to handle the request. Essentially, this first checks the database for the userid, and if it exists, updates the corresponding user with the given DTO. This then returns the updated user on success.
4. Added input validation to the UpdateUserDTO

### ✔️ Verification

 Created an exampleUser (lets call him Bob) to simulate endpoint callers. Created a local MongoDB database "c4cOpsTest" with multiple users and updated Bob to check for expected behavior in the following scenarios:
- change Bob's status to Admin. Given an existing userID and valid shema body-> 200 OK + update any userID's associated User and return the updated user 
- change Bob's status to Applicant. Given a userID that is not Bob's -> 400 Bad Request
- change Bob's status to Applicant/ALUMNI. Given a UserId with Applicant Status -> 400 BAD REQUEST


### 🏕️ (Optional) Future Work / Notes

- maybe add an error for when no parameters given
- Update UpdateUserDTO input validation with Regex to check whether links are in valid format
